### PR TITLE
Small repairs

### PR DESCRIPTION
**Pull request by Koud-Wind**
## 📝 Description
Better death message, removal of deceleration during knockback, repair of weapon freezing caused by sneak during reload in creative mode, sorting of configuration options

## 🚦 Testing 
 - [x] Client
 - [x] Server

## 🖼️ Screenshots/Recordings

image

image

https://github.com/Cubed-Development/Modern-Warfare-Cubed/assets/123817318/21c3f173-cd55-4c63-a3e3-b34bb918c7dc

https://github.com/Cubed-Development/Modern-Warfare-Cubed/assets/123817318/69758742-f2f3-4fd5-8654-3b317761f07e

## ⏮️ Backwards Compatibility 
Increase compatibility

### CODE REVIEW

**Comment by Desoroxxx** [comment]
> removal of deceleration during knockback

So you removed the knockback working fine?

**Comment by Koud-Wind** [comment]
> So you removed the knockback working fine?
Yeah, its working is fine, it can also cause a small amount of knockback

File: src/main/java/com/paneedah/weaponlib/WeaponReloadAspect.java [hunk 1]
```diff
@@ -564,24 +564,25 @@ private void processActualCompoundPermit(CompoundPermit p, PlayerWeaponInstance
...
+        // Mag list is empty
         if (compatibleMagazines.isEmpty())
             return;
 
-        ItemStack magazineStack = MWCUtil.consumeItemsFromPlayerInventory(compatibleMagazines, (stack1, stack2) -> Integer.compare(Tags.getAmmo(stack1), Tags.getAmmo(stack2)), player);
+        ItemStack magazineStack = MWCUtil.consumeItemsFromPlayerInventory(compatibleMagazines, Comparator.comparingInt(Tags::getAmmo), player);
 
-        if (magazineStack == null)
+        if (magazineStack == null) {
+            // Cancel unload when magazine cannot be found in inventory
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/WeaponReloadAspect.java:581]: This just seems like a bug waiting to happen.

File: src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java [hunk 2]
```diff
@@ -82,12 +88,12 @@ protected void onImpact(final RayTraceResult position) {
             Explosion.createServerSideExplosion(weapon.getModContext(), world, this.getThrower(), this, position.hitVec.x, position.hitVec.y, position.hitVec.z, explosionRadius, false, true, isDestroyingBlocks, explosionParticleAgeCoefficient, smokeParticleAgeCoefficient, explosionParticleScaleCoefficient, smokeParticleScaleCoefficient, weapon.getModContext().getRegisteredTexture(explosionParticleTextureId), weapon.getModContext().getRegisteredTexture(smokeParticleTextureId), weapon.getModContext().getExplosionSound());
         } else if (position.entityHit != null) {
             if (this.getThrower() != null)
-                position.entityHit.attackEntityFrom(DamageSource.causeThrownDamage(this, this.getThrower()), damage);
+                position.entityHit.attackEntityFrom(new ProjectileDamageSource(this, this.getThrower()), damage);
             else
-                position.entityHit.attackEntityFrom(new DamageSource("thrown"), damage);
+                position.entityHit.attackEntityFrom(new DamageSource("arrow"), damage);
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java:93]: We already said we won't be using 'arrow' as the damage source. It's not supposed to be like this.

File: src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java [hunk 3]
```diff
@@ -82,12 +88,12 @@ protected void onImpact(final RayTraceResult position) {
             Explosion.createServerSideExplosion(weapon.getModContext(), world, this.getThrower(), this, position.hitVec.x, position.hitVec.y, position.hitVec.z, explosionRadius, false, true, isDestroyingBlocks, explosionParticleAgeCoefficient, smokeParticleAgeCoefficient, explosionParticleScaleCoefficient, smokeParticleScaleCoefficient, weapon.getModContext().getRegisteredTexture(explosionParticleTextureId), weapon.getModContext().getRegisteredTexture(smokeParticleTextureId), weapon.getModContext().getExplosionSound());
         } else if (position.entityHit != null) {
             if (this.getThrower() != null)
-                position.entityHit.attackEntityFrom(DamageSource.causeThrownDamage(this, this.getThrower()), damage);
+                position.entityHit.attackEntityFrom(new ProjectileDamageSource(this, this.getThrower()), damage);
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java:91]: Please stop making the damage sourse a vanilla arrow. That's not correct.

File: src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java [hunk 4]
```diff
@@ -183,4 +189,19 @@ public void shoot(final double x, final double y, final double z, final float ve
     public Weapon getWeapon() {
         return weapon;
     }
+
+    public static class ProjectileDamageSource extends EntityDamageSourceIndirect {
+        public ProjectileDamageSource(Entity projectile, @Nonnull Entity shooter) {
+            super("arrow", projectile, shooter);
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java:195]: Once again - all of this is trying to use an arrow as the projectile damage source.
Please stop.

File: src/main/java/com/paneedah/weaponlib/animation/movement/WeaponRotationHandler.java [hunk 5]
```diff
@@ -51,7 +51,8 @@ public class WeaponRotationHandler {
 	
 	
 	private static final float SWAY_AMPLITUDE_NORMAL = 0.15f;
-	private static final float SWAY_AMPLITUDE_OPTICAL_ADS = 0.03f;
+	private static final float SWAY_AMPLITUDE_OPTICAL_ADS = 0.10f;
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/animation/movement/WeaponRotationHandler.java:54]: Are you crazy? 3x more sway??

File: src/main/java/com/paneedah/weaponlib/core/WeaponlibClassTransformer.java [hunk 6]
```diff
@@ -757,8 +774,8 @@ else if (layerHeldItemClassInfo != null
                     && entityLivingBaseClassInfo.methodMatches("attackEntityFrom", "(Lnet/minecraft/util/DamageSource;F)Z", classname, name, desc)) {
                 return new AttackEntityFromMethodVisitor(cv.visitMethod(access, name, desc, signature, exceptions));
             } else if (entityLivingBaseClassInfo != null
-                    && entityLivingBaseClassInfo.methodMatches("attackEntityFrom", "(Lnet/minecraft/util/DamageSource;F)Z", classname, name, desc)) {
-                return new AttackEntityFromMethodVisitor(cv.visitMethod(access, name, desc, signature, exceptions));
+                    && entityLivingBaseClassInfo.methodMatches("knockBack", "(Lnet/minecraft/entity/Entity;FDD)V", classname, name, desc)) {
+                return new KnockBackMethodVisitor(cv.visitMethod(access, name, desc, signature, exceptions));
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/core/WeaponlibClassTransformer.java:778]: Why?

File: src/main/java/com/paneedah/weaponlib/perspective/OpticalScopePerspective.java [hunk 7]
```diff
@@ -67,7 +67,7 @@ private static boolean isAimingState(RenderableState renderableState) {
     public void update(TickEvent.RenderTickEvent event) {
     	
          PlayerWeaponInstance instance = modContext.getMainHeldWeapon();
-        if(instance != null && instance.isAimed()) {
```
Comment (Paneedah) [inline src/main/java/com/paneedah/weaponlib/perspective/OpticalScopePerspective.java:70]: This causes the whole world to be rendered twice. That is not a good idea.

**Review by Paneedah — CHANGES_REQUESTED** [review]

**Comment by Paneedah** [comment]
Please also follow the PR template.

File: src/main/java/com/paneedah/weaponlib/WeaponReloadAspect.java (hunk 1 above)
Comment (Koud-Wind) [reply src/main/java/com/paneedah/weaponlib/WeaponReloadAspect.java:581]: Is there any problem with it?

File: src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java (hunk 3 above)
Comment (Koud-Wind) [reply src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java:91]: The source of damage is not an arrow, it's just that the type of death message is an arrow

File: src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java (hunk 4 above)
Comment (Koud-Wind) [reply src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java:195]: The source of damage is not an arrow, it's just that the type of death message is an arrow

File: src/main/java/com/paneedah/weaponlib/core/WeaponlibClassTransformer.java (hunk 6 above)
Comment (Koud-Wind) [reply src/main/java/com/paneedah/weaponlib/core/WeaponlibClassTransformer.java:778]: I see that the knockback algorithm will halve the motion of the original entity

File: src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java (hunk 2 above)
Comment (Koud-Wind) [reply src/main/java/com/paneedah/weaponlib/WeaponSpawnEntity.java:93]: The source of damage is not an arrow, it's just that the type of death message is an arrow

**Pull request closed without merging**